src/sink/ws.rs

```diff
@@ -42,138 +40,273 @@ struct WsMessageMeta {
-                    &sink_url,
-                    url, e
-                );
-                // TODO avoid these clones for non linked-transport usecase
-                offramp_tx.send(WsResult::Fail(id.clone(), meta)).await?;
-                if has_link {
-                    offramp_tx
-                        .send(WsResult::Response(
-                            id,
-                            Box::new(Err(Error::from(format!(
-                                "[Sink::{}] Error sending event to server {}: {}",
-                                &sink_url, url, e
-                            )))),
-                        ))
-                        .await?;
+        'recv_loop: while let Ok(ConnectionTaskMsg::SendEvent(
+            ids,
+            message_meta,
+            op_meta,
+            ingest_ns,
+            data,
+        )) = rx.recv().await
+        {
+            match event_to_message(
+                &codec,
+                &mut postprocessors,
+                ingest_ns,
+                &data,
+                message_meta.binary,
+            ) {
+                Ok(iter) => {
+                    for msg_result in iter {
+                        match msg_result {
+                            Ok(msg) => {
+                                match ws_stream.send(msg).await {
+                                    Ok(_) => {
+                                        let mut e = Event::cb_ack(nanotime(), ids.clone());
+                                        e.op_meta = op_meta.clone();
+                                        reply_tx.send(sink::Reply::Insight(e)).await?;
+                                    }
+                                    Err(e) => {
+                                        let e = format!(
+                                            "Error sending event to server {}: {}.",
+                                            &url, e
+                                        );
+                                        handle_error_with_fail(
+                                            &sink_url,
+                                            &e,
+                                            &reply_tx,
+                                            &ids,
+                                            &event_origin_url,
+                                            op_meta.clone(),
+                                        )
+                                        .await?;
+
+                                        // close connection explicitly
+                                        ws_stream.close(None).await?;
+                                        connection_lifecycle_tx
+                                            .send(WsConnectionMsg::Disconnected(url.clone()))
+                                            .await?;
+                                        break 'recv_loop; // exit recv loop in order to reconnect
```

Comment:
Wouldn't we need to fail the remaining messages at this point?

Reply:
Good catch, i am going to send only one ack if all messages coming from an event succeeded, and a fail otherwise as soon as it happens. We might have send some messages, before we failed, but at least there is only 1 ack or fail flowing backwards per event.

Reply:
👍 it's in #583 and #584